File: tests/FSharpx.DataStructures.Tests/NonEmptyListTests.fs
module FSharpx.NonEmptyListTests

open FSharpx.Collections
open NUnit.Framework
open FsCheck
open FsCheck.NUnit

type NonEmptyListGen =
    static member NonEmptyList() =
        gen.Return NonEmptyList.create 
        |> Gen.ap Arb.generate
        |> Gen.ap (Arb.generate |> Gen.suchThat (fun l -> l.Length < 10))
        |> Arb.fromGen

let registerGen = lazy (Arb.register<NonEmptyListGen>() |> ignore)

[<Test>]
let ``functor laws``() =
    registerGen.Force()
    let n = sprintf "NonEmptyList : functor %s"
    let map = NonEmptyList.map
    fsCheck (n "preserves identity") <| 
        fun value -> map id value = value
    fsCheck (n "preserves composition") <|
        fun f g value -> map (f << g) value = (map f << map g) value

[<Test>]
let ``monad laws``() =
    registerGen.Force()
    let ret (x: int) = NonEmptyList.singleton x
    let n = sprintf "NonEmptyList : monad %s"
    let inline (>>=) m f = NonEmptyList.collect f m
    fsCheck "left identity" <| 
        fun f a -> ret a >>= f = f a
    fsCheck "right identity" <| 
        fun x -> x >>= ret = x
    fsCheck "associativity" <| 
        fun f g v ->
            let a = (v >>= f) >>= g
            let b = v >>= (fun x -> f x >>= g)
            a = b

let fsCheck t = fsCheck "" t

[<Test>]
let ``toList gives non-empty list``() =
    fsCheck (fun nel -> NonEmptyList.toList nel |> List.length > 0)

[<Test>]
let ``toArray gives non-empty array``() =
    fsCheck (fun nel -> NonEmptyList.toArray nel |> Array.length > 0)

[<Test>]
let ``toList is same length as non-empty list`` () =
    fsCheck (fun nel -> NonEmptyList.toList nel |> List.length = nel.Length)

[<Test>]
let ``toArray is same length as non-empty list`` () =
    fsCheck (fun nel -> NonEmptyList.toArray nel |> Array.length = nel.Length)

[<Test>]
let ``reverse . reverse = id`` () =
    fsCheck (fun nel -> (NonEmptyList.rev << NonEmptyList.rev) nel = nel)

[<Test>]
let ``last . reverse = head``() =
    fsCheck (fun nel -> (NonEmptyList.last << NonEmptyList.rev) nel = NonEmptyList.head nel)

[<Test>]
let ``head . reverse = last``() =
    fsCheck (fun nel -> (NonEmptyList.head << NonEmptyList.rev) nel = NonEmptyList.last nel)

[<Test>]
let ``last is last and never fails``() =
    fsCheck <| fun nel ->
        let actualLast = NonEmptyList.last nel
        let expectedLast = 
            let l = NonEmptyList.toList nel
            l.[l.Length-1]
        expectedLast = actualLast

[<Test>]
let ``append has combined length``() =
    fsCheck <| fun a b ->
        let c = NonEmptyList.append a b
        c.Length = a.Length + b.Length

[<Test>]
let reduce() =
    fsCheck <| fun nel ->
        let actual = NonEmptyList.reduce (+) nel
        let expected = nel |> NonEmptyList.toList |> List.sum
        expected = actual

File: tests/FSharpx.CSharpTests/NonEmptyListTests.cs
using Microsoft.FSharp.Collections;
using NUnit.Framework;
using FSharpx.Collections;


namespace FSharpx.CSharpTests {
    [TestFixture]
    public class NonEmptyListTests {
        [Test]
        public void Create() {
            NonEmptyList<int> x = NonEmptyList.Create(1, FSharpList.Create(2, 3, 4));
            NonEmptyList<int> y = NonEmptyList.Create(1, 2, 3, 4);
            Assert.AreEqual(4, x.Length);
            Assert.AreEqual(x,y);
        }

        [Test]
        public void Singleton() {
            NonEmptyList<int> x = NonEmptyList.Singleton(1);
            Assert.AreEqual(1, x.Length);
        }

        [Test]
        public void HeadTail() {
            NonEmptyList<int> x = NonEmptyList.Singleton(1);
            Assert.AreEqual(1, x.Head);
            Assert.AreEqual(FSharpList<int>.Empty, x.Tail);
        }

        [Test]
        public void ToFSharpList() {
            NonEmptyList<int> x = NonEmptyList.Singleton(1);
            FSharpList<int> list = x.ToFSharpList();
            Assert.AreEqual(FSharpList.Create(1), list);
        }

        [Test]
        public void ToArray() {
            NonEmptyList<int> x = NonEmptyList.Singleton(1);
            int[] a = x.ToArray();
            CollectionAssert.AreEqual(new[] {1}, a);
        }

        [Test]
        public void Concat_FSharpList() {
            var x = NonEmptyList.Singleton(1);
            x = NonEmptyList.Concat(x, FSharpList.Create(2, 3, 4));
            CollectionAssert.AreEqual(new[] {1,2,3,4}, x);
        }

        [Test]
        public void Concat_NonEmptyList() {
            var x = NonEmptyList.Singleton(1);
            x = NonEmptyList.Concat(x, NonEmptyList.Singleton(2));
            CollectionAssert.AreEqual(new[] { 1, 2 }, x);
        }

        [Test]
        public void Concat_FSharpList_extension() {
            var x = NonEmptyList.Singleton(1);
            x = x.Concat(FSharpList.Create(2, 3, 4));
            CollectionAssert.AreEqual(new[] { 1, 2, 3, 4 }, x);
        }

        [Test]
        public void Concat_NonEmptyList_extension() {
            var x = NonEmptyList.Singleton(1);
            x = x.Concat(NonEmptyList.Singleton(2));
            CollectionAssert.AreEqual(new[] { 1, 2 }, x);
        }

        [Test]
        public void Cons() {
            var x = NonEmptyList.Singleton(1).Cons(2);
            CollectionAssert.AreEqual(new[] { 2, 1 }, x);
        }

        [Test]
        public void Reverse() {
            var x = NonEmptyList.Create(1, FSharpList.Create(2, 3, 4));
            x = x.Reverse();
            CollectionAssert.AreEqual(new[] { 4,3,2,1 }, x);
        }

        [Test]
        public void Select() {
            var list = NonEmptyList.Create(1, FSharpList.Create(2, 3, 4));
            list = list.Select(x => x + 2);
            CollectionAssert.AreEqual(new[] {3,4,5,6}, list);
            list = from x in list select x + 1;
            CollectionAssert.AreEqual(new[] { 4, 5, 6, 7 }, list);
        }
    }
}
